go-controller/pkg/ovn/ovn.go

```diff
@@ -150,24 +150,59 @@ func (oc *DefaultNetworkController) ensurePod(oldPod, pod *kapi.Pod, addPort boo
+			// No matter if a pod is ovn networked, or host networked, we still need to check for exgw
+			// annotations. If the pod is ovn networked and is in update reschedule, addLogicalPort will take
+			// care of updating the exgw updates
+			if err := oc.deletePodExternalGW(oldPod); err != nil {
+				return fmt.Errorf("ensurePod failed %s/%s: %w", pod.Namespace, pod.Name, err)
+			}
 		}
-	}
 
-	if !util.PodWantsHostNetwork(pod) && addPort {
-		if err := oc.addLogicalPort(pod); err != nil {
-			return fmt.Errorf("addLogicalPort failed for %s/%s: %w", pod.Namespace, pod.Name, err)
+		if !util.PodWantsHostNetwork(pod) && addPort {
+			if err := oc.addLogicalPort(pod); err != nil {
+				return fmt.Errorf("addLogicalPort failed for %s/%s: %w", pod.Namespace, pod.Name, err)
+			}
+		} else {
+			// either pod is host-networked or its an update for a normal pod (addPort=false case)
+			if oldPod == nil || exGatewayAnnotationsChanged(oldPod, pod) || networkStatusAnnotationsChanged(oldPod, pod) {
+				if err := oc.addPodExternalGW(pod); err != nil {
+					return fmt.Errorf("addPodExternalGW failed for %s/%s: %w", pod.Namespace, pod.Name, err)
+				}
+			}
 		}
 	} else {
-		// either pod is host-networked or its an update for a normal pod (addPort=false case)
-		if oldPod == nil || exGatewayAnnotationsChanged(oldPod, pod) || networkStatusAnnotationsChanged(oldPod, pod) {
-			if err := oc.addPodExternalGW(pod); err != nil {
-				return fmt.Errorf("addPodExternalGW failed for %s/%s: %w", pod.Namespace, pod.Name, err)
+		// Track remote ovn networked pods too, for network policy, and egress gw.
+		if len(pod.Status.PodIPs) >= 1 {
+			var ips []*net.IPNet
+			for _, podIp := range pod.Status.PodIPs {
+				ip := net.ParseIP(podIp.IP)
+				if ip != nil {
+					ips = append(ips, &net.IPNet{IP: ip})
+				}
+
+			}
+			if (addPort || (oldPod != nil && len(pod.Status.PodIPs) != len(oldPod.Status.PodIPs))) && !util.PodWantsHostNetwork(pod) {
+				if err := oc.addRemotePodToNamespace(pod.Namespace, ips); err != nil {
+					return fmt.Errorf("failed to add non-local pod %s/%s to namespace: %v", pod.Namespace, pod.Name, err)
+				}
+			}
+
+			//FIXME: Update comments & reduce code duplication.
+			// check if this pod is serving as an external GW
+			if oldPod != nil && (exGatewayAnnotationsChanged(oldPod, pod) || networkStatusAnnotationsChanged(oldPod, pod)) {
+				// No matter if a pod is ovn networked, or host networked, we still need to check for exgw
+				// annotations. If the pod is ovn networked and is in update reschedule, addLogicalPort will take
+				// care of updating the exgw updates
+				_ = oc.deletePodExternalGW(oldPod)
+			}
+
+			// either pod is host-networked or its an update for a normal pod (addPort=false case)
+			if oldPod == nil || exGatewayAnnotationsChanged(oldPod, pod) || networkStatusAnnotationsChanged(oldPod, pod) {
+				if err := oc.addPodExternalGW(pod); err != nil {
```

Comment:
I am confused about which parts of a local pod setup we need for non-local pods. Here we do `addPodExternalGW`, but then in `DeleteResource` we don't call `deletePodExternalGW` for non-local pods.
It would be nice to have some general comment about which actions are required for non-local pods (maybe it is already present somewhere, but I didn't find it?)

Reply:
Thanks for finding this.  Have fixed in the latest push.